src/v/cluster/controller.cc

```diff
@@ -537,10 +542,29 @@ ss::future<> controller::cluster_creation_hook(cluster_discovery& discovery) {
         vassert(
           config::node().seed_servers().empty(), "We are not on the root node");
 
-        if (
-          _raft0->last_visible_index() > model::offset{}
-          || _raft0->config().brokers().size() > 1) {
+        if (controller_log_exists()) {
             // The controller log has already been written to
+            if (!_storage.local().get_cluster_uuid()) { // HERE: ctrllog done?
+                // No cluster UUID - we are upgrading an existing cluster
+                vlog(
+                  clusterlog.info,
+                  "Cluster will be upgraded once all nodes are upgraded");
+                ssx::background
+                  = _feature_table.local()
+                      .await_feature(
+                        features::feature::seeds_driven_bootstrap_capable,
+                        _as.local())
+                      .then([this] { return create_cluster(); })
+                      .then([] {
+                          vlog(clusterlog.info, "Cluster has been upgraded");
+                      })
+                      .handle_exception([](const std::exception_ptr e) {
+                          vlog(
+                            clusterlog.warn,
+                            "Error upgrading the cluster. {}",
+                            e);
```

Comment:
I think this needs to consider whether or not the local node is a leader.

```
future<> generate_cluster_uuid() {
    co_await _feature_table.local().await_feature(features::feature::seeds_driven_bootstrap, _as.local());
    while (!_storage.local().get_cluster_uuid() && !_raft0->is_leader()) {
        co_await ss::sleep(100ms);
    }
    if (_storage.local().get_cluster_uuid()) {
        co_return;
    }
    create_cluster();
}
```
Might be worth wrapping the bottom half in some `create_or_wait_for_cluster()` or somesuch, since that could be reused below.

Reply:
Also I thought we'd only create a UUID, not everything else?

Reply:
Both points are valid. I've missed the fact that the fiber goes off _before_ we check whether we are the leader.